File: BACK/src/Modules/Encounters/Explorer.Encounters.Core/UseCases/KeypointEncounterService.cs
using AutoMapper;
using Explorer.BuildingBlocks.Core.UseCases;
using Explorer.Encounters.API.Dtos;
using Explorer.Encounters.API.Public;
using Explorer.Encounters.Core.Domain;
using Explorer.Encounters.Core.Domain.Enums;
using Explorer.Encounters.Core.Domain.RepositoryInterfaces;
using Explorer.Tours.Core.Domain;
using Explorer.Stakeholders.API.Dtos;
using Explorer.Stakeholders.API.Internal;
using Explorer.Stakeholders.Core.Domain;

using FluentResults;

namespace Explorer.Encounters.Core.UseCases
{
     public class KeypointEncounterService : CrudService<KeypointEncounterDto, KeypointEncounter>, IKeypointEncounterService
    {
        protected IKeypointEncounterRepository _keypointEncounterRepository;
        protected IEncounterRepository _encounterRepository;
        protected IInternalUserService _userService;
        protected IInternalProfileService _profileService;

        public KeypointEncounterService(IKeypointEncounterRepository keypointEncounterRepository, IEncounterRepository encounterRepository,
             IInternalUserService userService, IInternalProfileService profileService, IMapper mapper) : base(keypointEncounterRepository, mapper)
        {
            _keypointEncounterRepository = keypointEncounterRepository;
            _encounterRepository = encounterRepository;
            _userService = userService;
            _profileService = profileService;
        }
        public Result<PagedResult<KeypointEncounterDto>> GetPagedByKeypoint(int page, int pageSize, long keypointId)
        {
            var result = _keypointEncounterRepository.GetPagedByKeypoint(page, page, keypointId);
            return MapToDto(result);
        }
        public override Result<KeypointEncounterDto> Create(KeypointEncounterDto keypointEncounter)
        {
            try
            {
                keypointEncounter.Encounter.ApprovalStatus = EncounterApprovalStatus.SYSTEM_APPROVED.ToString();
                var result = _keypointEncounterRepository.Create(MapToDomain(keypointEncounter));
                return MapToDto(result);
            }
            catch (Exception e)
            {
                return Result.Fail(FailureCode.InvalidArgument).WithError(e.Message);
            }
        }

        public Result Delete(int keypointEncounterId)
        {
            try
            {
                var keypointEncounter = _keypointEncounterRepository.Get(keypointEncounterId);

                if(keypointEncounter == null) return Result.Fail(FailureCode.NotFound).WithError("Encounter is not find");
                
                _keypointEncounterRepository.Delete(keypointEncounterId);
                _encounterRepository.Delete(keypointEncounter.Encounter.Id);
                return Result.Ok();
            }
            catch (KeyNotFoundException e)
            {
                return Result.Fail(FailureCode.NotFound).WithError(e.Message);
            }
        }

        public Result UpdateEncountersLocation(LocationDto location, int keypointId)
        {
            try
            {
                var keypointEncounters = _keypointEncounterRepository.GetAllByKeypoint(keypointId);
                foreach (var keypointEncounter in keypointEncounters)
                {
                    keypointEncounter.Encounter.UpdateLocation(location.Latitude, location.Longitude);
                    _encounterRepository.Update(keypointEncounter.Encounter);
                }
                return Result.Ok();
            }
            catch (KeyNotFoundException e)
            {
                return Result.Fail(FailureCode.Conflict).WithError(e.Message);
            }
        }
        public Result DeleteKeypointEncounters(int keypointId)
        {
            var keypointEncounters = _keypointEncounterRepository.GetAllByKeypoint(keypointId);
                foreach (var keypointEncounter in keypointEncounters)
                {
                    Delete((int)keypointEncounter.Id);
                }
                
            return Result.Ok();
        }
    }
}

File: ENCOUNTERS-MS/service/KeypointEncounterService.go
package service

import (
	"ENCOUNTERS-MS/model"
	"ENCOUNTERS-MS/repo"
	"fmt"
)

type KeypointEncounterService struct {
	KeypointEncounterRepo *repo.KeypointEncounterRepository
}

func (service *KeypointEncounterService) GetPagedByKeypoint(keypointId string) ([]*model.KeypointEncounter, error) {
	result, err := service.KeypointEncounterRepo.GetPagedByKeypoint(keypointId)
	if err != nil {
		return nil, fmt.Errorf("Error fetching paged keypoint encounters: %s", err)
	}
	return result, nil
}

func (service *KeypointEncounterService) Create(keypointEncounter *model.KeypointEncounter) (*model.KeypointEncounter, error) {
	keypointEncounter.Encounter.ApprovalStatus = model.ApprovalSystemApproved
	result, err := service.KeypointEncounterRepo.Create(keypointEncounter)
	if err != nil {
		return nil, fmt.Errorf("Error creating keypoint encounter: %s", err)
	}
	return result, nil
}

func (service *KeypointEncounterService) Delete(keypointEncounterId string) error {
	keypointEncounter, err := service.KeypointEncounterRepo.GetById(keypointEncounterId)
	if err != nil {
		return fmt.Errorf("Error finding keypoint encounter: %s", err)
	}
	if keypointEncounter == nil {
		return fmt.Errorf("Encounter is not found")
	}

	if err := service.KeypointEncounterRepo.Delete(keypointEncounterId); err != nil {
		return fmt.Errorf("Error deleting keypoint encounter: %s", err)
	}

	return nil
}

func (service *KeypointEncounterService) Update(keypointEncounter *model.KeypointEncounter) (*model.KeypointEncounter, error) {
	result, err := service.KeypointEncounterRepo.Update(keypointEncounter)
	if err != nil {
		return nil, fmt.Errorf("Error updating keypoint encounter: %s", err)
	}
	return result, nil
}
